Unable to download from new bucket temporarily

**Detailed Description**
When creating a new bucket and re-pointing the importer in staging to it, it was unable to download from the bucket with 307 errors. This is most likely a transient error as the new bucket rolls out to all regions. But it indicates we aren't configuring the application properly and using a non-region specific region endpoint.

**Actual Behavior**
Steps to reproduce the behavior:
1. Stop application
2. Create new bucket
3. Update app config
4. Start app

```
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]: 2020-01-22 20:07:27,948 ERROR [pool-7-thread-13] c.h.m.i.d.b.AccountBalancesDownloader Error downloading signature files for node 0.0.5 after 32.81 ms
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]: java.util.concurrent.ExecutionException: software.amazon.awssdk.services.s3.model.S3Exception: Please re-send this request to the specified temporary endpoint. Continue to use the original request endpoint for future requests. (Service: S3, Status Code: 307, Request ID: 612F8F5FA4147D2D)
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395) ~[?:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999) ~[?:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at com.hedera.mirror.importer.downloader.Downloader.lambda$downloadSigFiles$1(Downloader.java:161) ~[classes!/:0.5.2]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at java.lang.Thread.run(Thread.java:834) [?:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]: Caused by: software.amazon.awssdk.services.s3.model.S3Exception: Please re-send this request to the specified temporary endpoint. Continue to use the original request endpoint for future requests. (Service: S3, Status Code: 307, Request ID: 612F8F5FA4147D2D)
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at software.amazon.awssdk.services.s3.model.S3Exception$BuilderImpl.build(S3Exception.java:95) ~[s3-2.10.6.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at software.amazon.awssdk.services.s3.model.S3Exception$BuilderImpl.build(S3Exception.java:55) ~[s3-2.10.6.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at software.amazon.awssdk.protocols.query.unmarshall.AwsXmlErrorProtocolUnmarshaller.handle(AwsXmlErrorProtocolUnmarshaller.java:127) ~[aws-query-protocol-2.10.6.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at software.amazon.awssdk.protocols.query.unmarshall.AwsXmlErrorProtocolUnmarshaller.handle(AwsXmlErrorProtocolUnmarshaller.java:86) ~[aws-query-protocol-2.10.6.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler.lambda$prepare$0(AsyncResponseHandler.java:88) ~[sdk-core-2.10.6.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072) ~[?:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073) ~[?:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler$BaosSubscriber.onComplete(AsyncResponseHandler.java:129) ~[sdk-core-2.10.6.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.runAndLogError(ResponseHandler.java:171) ~[netty-nio-client-2.10.6.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.access$500(ResponseHandler.java:68) ~[netty-nio-client-2.10.6.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler$PublisherAdapter$1.onComplete(ResponseHandler.java:287) ~[netty-nio-client-2.10.6.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at com.typesafe.netty.HandlerPublisher.publishMessage(HandlerPublisher.java:362) ~[netty-reactive-streams-2.0.3.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at com.typesafe.netty.HandlerPublisher.flushBuffer(HandlerPublisher.java:304) ~[netty-reactive-streams-2.0.3.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at com.typesafe.netty.HandlerPublisher.receivedDemand(HandlerPublisher.java:258) ~[netty-reactive-streams-2.0.3.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at com.typesafe.netty.HandlerPublisher.access$200(HandlerPublisher.java:41) ~[netty-reactive-streams-2.0.3.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at com.typesafe.netty.HandlerPublisher$ChannelSubscription$1.run(HandlerPublisher.java:452) ~[netty-reactive-streams-2.0.3.jar!/:?]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) ~[netty-common-4.1.43.Final.jar!/:4.1.43.Final]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:510) ~[netty-common-4.1.43.Final.jar!/:4.1.43.Final]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:518) ~[netty-transport-4.1.43.Final.jar!/:4.1.43.Final]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at io.netty.util.concurrent.SingleThreadEventExecutor$6.run(SingleThreadEventExecutor.java:1050) ~[netty-common-4.1.43.Final.jar!/:4.1.43.Final]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.43.Final.jar!/:4.1.43.Final]
Jan 22 20:07:27 mirrornode-staging1-1 java[11343]:         ... 1 more
```
**Expected Behavior**
Files downloaded

**Environment:**
- Java: OpenJDK 11
- OS: Ubuntu 18.04
- Version: v0.5.2

**Additional Context**

--- a/hedera-mirror-importer/pom.xml
+++ b/hedera-mirror-importer/pom.xml
@@ -163,7 +163,7 @@
             <dependency>
                 <groupId>software.amazon.awssdk</groupId>
                 <artifactId>bom</artifactId>
-                <version>2.10.6</version>
+                <version>2.10.54</version>
                 <type>pom</type>
                 <scope>import</scope>
             </dependency>

--- a/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/config/MirrorImporterConfiguration.java
+++ b/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/config/MirrorImporterConfiguration.java
@@ -22,6 +22,7 @@ package com.hedera.mirror.importer.config;
 
 import java.net.URI;
 import java.time.Duration;
+import lombok.RequiredArgsConstructor;
 import lombok.extern.log4j.Log4j2;
 import org.apache.commons.lang3.StringUtils;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
@@ -37,16 +38,20 @@ import software.amazon.awssdk.http.async.SdkAsyncHttpClient;
 import software.amazon.awssdk.http.nio.netty.NettyNioAsyncHttpClient;
 import software.amazon.awssdk.regions.Region;
 import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.S3AsyncClientBuilder;
 
 import com.hedera.mirror.importer.downloader.CommonDownloaderProperties;
 
 @Configuration
 @EnableAsync
 @Log4j2
+@RequiredArgsConstructor
 public class MirrorImporterConfiguration {
 
+    private final CommonDownloaderProperties downloaderProperties;
+
     @Bean
-    public S3AsyncClient s3AsyncClient(CommonDownloaderProperties downloaderProperties) {
+    public S3AsyncClient s3AsyncClient() {
         SdkAsyncHttpClient httpClient = NettyNioAsyncHttpClient.builder()
                 .maxConcurrency(downloaderProperties.getMaxConcurrency())
                 .connectionMaxIdleTime(Duration.ofSeconds(5))  // https://github.com/aws/aws-sdk-java-v2/issues/1122
@@ -57,10 +62,18 @@ public class MirrorImporterConfiguration {
 
         return S3AsyncClient.builder()
                 .credentialsProvider(awsCredentialsProvider(downloaderProperties))
-                .endpointOverride(URI.create(downloaderProperties.getCloudProvider().getEndpoint()))
                 .region(Region.of(downloaderProperties.getRegion()))
                 .httpClient(httpClient)
+                .applyMutation(this::overrideEndpoint)
                 .build();
+    }
+
+    private void overrideEndpoint(S3AsyncClientBuilder builder) {
+        // If Amazon S3, let the client generate the endpoint automatically based upon the region and bucket name
+        if (downloaderProperties.getCloudProvider() != CommonDownloaderProperties.CloudProvider.S3) {
+            URI endpoint = URI.create(downloaderProperties.getCloudProvider().getEndpoint());
+            builder.endpointOverride(endpoint);
+        }
     }
 
     @Bean

--- a/hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/downloader/AbstractDownloaderTest.java
+++ b/hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/downloader/AbstractDownloaderTest.java
@@ -96,7 +96,7 @@ public abstract class AbstractDownloaderTest {
         System.out.println("Before test: " + testInfo.getTestMethod().get().getName());
 
         initProperties();
-        s3AsyncClient = (new MirrorImporterConfiguration()).s3AsyncClient(commonDownloaderProperties);
+        s3AsyncClient = new MirrorImporterConfiguration(commonDownloaderProperties).s3AsyncClient();
         networkAddressBook = new NetworkAddressBook(mirrorProperties);
         downloader = getDownloader();
 

--- a/pom.xml
+++ b/pom.xml
@@ -48,7 +48,7 @@
     <parent>
         <groupId>org.springframework.boot</groupId>
         <artifactId>spring-boot-starter-parent</artifactId>
-        <version>2.2.2.RELEASE</version>
+        <version>2.2.4.RELEASE</version>
         <relativePath/> <!-- lookup parent from repository -->
     </parent>
 

